### ISSUE TITLE
distance_transform bug

### ISSUE THREAD

**Issue by michal-ed** [kornia/kornia#3279]
### Describe the bug

when generating a 3D distance map from a label volume ( size 128 128 64), in one of the classes - there are zeros in a rectangular region which is not connected to the edge of the structure in any way.
this does not occur when using scipy distance_transform_edt
 a slice from the masked volume
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/fe45f820-8388-4abe-bc4b-9ffb955319b7" />
sdf produced by scipy
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/caebda25-40c5-4e62-aa1a-0ba55d07d97d" />
result of the distance_transform - rectangular region does not exist in the original volume
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/17e0a979-1b5e-4685-9fc6-97c4926a93a2" />

### Reproduction steps

```bash
labels_one_hot = torch.nn.functional.one_hot(valid_labels.long(), num_classes=num_classes)  # [B, H, W, D, C]
        labels_one_hot = labels_one_hot.permute(4, 0, 1, 2, 3).float()  # [C, B, H, W, D]
        labels_one_hot *= mask.unsqueeze(0)  # mask: [B, H, W, D] -> [1, B, H, W, D]
        # Vectorized: stack all class masks as batch
        C, B, H, W, D = labels_one_hot.shape
        masks = labels_one_hot.permute(1, 0, 2, 3, 4).reshape(B * C, H, W, D)  # [B*C, H, W, D]
        dist_out = kornia.contrib.distance_transform(1 - masks)
        dist_in = kornia.contrib.distance_transform(masks)
        dist_out = dist_out.reshape(B, C, H, W, D).permute(1, 0, 2, 3, 4)  # [C, B, H, W, D]
        dist_in = dist_in.reshape(B, C, H, W, D).permute(1, 0, 2, 3, 4)    # [C, B, H, W, D]
        sdf = dist_out - dist_in  # [C, B, H, W, D]
        if squeeze_batch:
            sdf = sdf[:, 0]  # [C, H, W, D]
        if per_sample_normalize:
            # Normalize each class map separately
            max_val = sdf.flatten(1).abs().max(dim=1)[0]  # [C]
            max_val = max_val.view(-1, 1, 1, 1).clamp(min=1e-6)  # avoid div/0
            sdf = sdf / max_val
```

### Expected behavior

correct distance_transform to give similar results of scipy, or simply resolve the unwanted zeros that occur

### Environment

```shell
ubuntu 22
torch 2.7.1
kornia 0.8.1
python 3.12
```

**Comment by edgarriba** [comment]
are you open to send a fix ?

**Comment by Aryanpagaria** [comment]
@michal-ed  can i work on this issue

**Comment by michal-ed** [comment]
From my perspective you are welcome to. I do not have the time to solve this.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by Aryanpagaria** [comment]
ok ill work on it
thanks for your time

**Comment by aarush-github** [comment]
Hello @edgarriba, @michal-ed , I would like to work on this issue. I've investigated the "unwanted zeros" issue in `distance_transform` when used with 3D volumes.

From reproducing the bug locally (let me know if you would like to see the example code) and comparing kornia's output with `scipy.ndimage.distance_transform_edt`, it looks like kornia's `distance_transform` is not correctly handling 3D volumes.

**Potential Root Cause:** The current implementation relies on `filter2d` and assumes 4D inputs (B, C, H, W). When a 3D volume `(B, H, W, D)` is passed, it gets treated as a stack of independent 2D slices, without any interaction along the Z-axis. 
As a result, the transform cannot propagate distances across depth.

**Proposed Solution:** I suggest extending `distance_transform` to support 5D tensors.

Let me know if this approach makes sense to you. Happy to work on the issue further if assigned.

**Comment by edgarriba** [comment]
Feel free to give it a shot

**Comment by kirandharmar867-maker** [comment]
hi @michal-ed,
yesterday i alredy worked On this type issue,
if You say OK!, today i will fix it
Thank You.

**Comment by edgarriba** [comment]
> hi @michal-ed,
> yesterday i alredy worked On this type issue,
> if You say OK!, today i will fix it
> Thank You.

Please, no bots are allowed

**Comment by kirandharmar867-maker** [comment]
> > hi [@michal-ed](https://github.com/michal-ed),
> > yesterday i alredy worked On this type issue,
> > if You say OK!, today i will fix it
> > Thank You.
> 
> Please, no bots are allowed

Hello @edgarriba 
Understood.
I am a genuine contributor and I’ll work on this issue manually.
I will analyze the bug, implement the fix myself, and submit a proper PR with tests.

Thank you for the Clarification.

**Comment by edgarriba** [comment]
@kirandharmar867-maker what's you proposal ?

**Comment by AdityaKumarSethia** [comment]
# Proposal: Extension to N-Dimensional Distance Transform

Hi, @edgarriba @michal-ed, regarding the `distance_transform` 3D support:

**Analysis:** The current implementation relies on `filter2d` and a hardcoded 2D kernel construction (`torch.hypot`). When a 5D tensor (B, C, D, H, W) is passed, it fails because:
1. The shape validation raises a `ValueError`.
2. The kernel logic ignores the Depth (Z) dimension, preventing distance propagation across slices.

**Proposed Solution:** I propose extending `kornia.contrib.distance_transform` to support both 2D and 3D inputs dynamically:
1. Dynamic Shape Check: Allow inputs of shape `(B, C, H, W)` OR `(B, C, D, H, W)`.
2. 3D Kernel Generation:
    + Detect if input is 3D.
    + Use `create_meshgrid3d` to generate a volumetric kernel.
    + Compute Euclidean kernel weights using $\sqrt{x^2 + y^2 + z^2}$.
3. Convolution: Switch between `filter2d` and `filter3d` based on input dimensionality.

**Verification Plan:** I will compare the output against `scipy.ndimage.distance_transform_edt` (as suggested by the OP) to ensure numerical parity for 3D volumes.I will add a specific test case for Z-axis propagation to prevent regressions.I have a working prototype locally and will open a PR shortly to address this.

**Comment by aarush-github** [comment]
Hey @edgarriba, the CI is failing in the `tutorials` job (connected_components.ipynb), but it seems to be an existing issue with kornia_rs loading `cells_binary.png` and unrelated to my 3D Distance Transform changes.
Please let me know if you need any adjustments from my end. :)

**Comment by edgarriba** [comment]
that's a known issue, no worries. Added few more comments

**Comment by kirandharmar867-maker** [comment]
> [@kirandharmar867-maker](https://github.com/kirandharmar867-maker) what's you proposal ?

@edgarriba,
why, what is problem.
